File: src/offer/Questions22.java
package offer;

import offer.LinkedList.Node;

//链表的倒数第K个节点

public class Questions22 {
	
	public static void main(String[] arg0) {
		Node<String> node1 = new Node<>();
		node1.set("a");
		Node<String> node2 = new Node<>();
		node2.set("b");
		node1.setNext(node2);
		Node<String> node3 = new Node<>();
		node3.set("c");
		node2.setNext(node3);
		Node<String> node4 = new Node<>();
		node4.set("d");
		node3.setNext(node4);
		Node<String> node5 = new Node<>();
		node5.set("e");
		node4.setNext(node5);
		Node<String> node6 = new Node<>();
		node6.set("f");
		node5.setNext(node6);
		System.out.println(nodeK(node1, 2).get());
		System.out.println(nodeK(node1, 3).get());
		System.out.println(nodeK(node1, 4).get());
		Node<String> node = null;
		System.out.println(nodeK(node, 3).get());
		System.out.println(nodeK(node1, 0).get());
		System.out.println(nodeK(node1, 7).get());
	}
	
	public static <T> Node<T> nodeK(Node<T> first, int k) {
		if (first == null) {
			throw new NullPointerException("输入的链表头节点为空");
		}
		if (k <= 0) {
			throw new IllegalArgumentException("输入的数字k必须大于0");
		}
		Node<T> one = first;
		Node<T> two = first;
		for (int i = 0; i < k - 1; i++) {
			one = one.getNext();
			if (one == null) {
				throw new IllegalArgumentException("输入的链表没有最多没有k项");
			}
		}
		while (one.getNext() != null) {
			one = one.getNext();
			two = two.getNext();
		}
		/*for (int i = 1; one.getNext() != null; i++) {
			one = one.getNext();
			if (i > k - 1) {
				two = two.getNext();
			}
		}*/
		return two;
	}

}

File: src/macosMain/kotlin/com.qiaoyuang.algorithm/Questions22.kt
package com.qiaoyuang.algorithm

fun main() {
	val node1 = Node("a")
	val node2 = Node("b")
	node1.next = node2
	val node3 = Node("c")
	node2.next = node3
	val node4 = Node("d")
	node3.next = node4
	val node5 = Node("e")
	node4.next = node5
	val node6 = Node("f")
	node5.next = node6
	
	println("倒数第3个节点为${node1.findKthToTail(3)}")
	println("倒数第6个节点为${node1.findKthToTail(6)}")
	println()
	println("链表的中间节点为：${node1.findMidNode()}")
}

/**
 * 链表的倒数第K个节点
 */
fun <T> Node<T>.findKthToTail(k: Int): T {
	require(k > 0) { "输入的数字必须大于 0" }
	var priorCount = 1
	var laterCount = 1
	var prior: Node<T>? = this
	var later: Node<T>? = this
	while (prior?.next != null) {
		prior = prior.next
		priorCount++
		if (priorCount > k) {
			later = later?.next
			laterCount++
		}
	}
	require(priorCount - laterCount != k - 1) { "输入的数字k不能大于链表总长" }
	return later!!.t!!
}

/*
 * 相关题目：求链表的中间节点
 */
fun <T> Node<T>.findMidNode(): T {
	var prior: Node<T>? = this
	var later: Node<T>? = this
	while (prior?.next != null) {
		prior = prior.next
		prior = prior?.next
		later = later?.next
	}
	return later!!.t!!
}